### ISSUE TITLE
[Research] Evaluate RADAR-inspired difficulty and model-budget routing

### ISSUE THREAD

**Issue by Xunzhuo** [vllm-project/semantic-router#1166]
## Summary

Evaluate whether a RADAR-inspired difficulty/ability model improves routing among approved model-and-reasoning-budget configurations.

## Research source

[RADAR: Reasoning–Ability and Difficulty-Aware Routing for Reasoning LLMs](https://openreview.net/forum?id=9k1oXUtrhO) was published as a MATH-AI 2025 workshop poster. The paper models query difficulty and model-budget ability with an item-response-theory-inspired formulation and reports experiments on eight reasoning benchmarks, including out-of-distribution evaluation.

Those are paper results, not evidence that the method improves this repository's workloads. The prior issue text's “30–50% cost reduction,” exact onboarding sample count, and benchmark list were not supported as repository acceptance claims and are removed.

## Audited current-main baseline (2026-07-15)

- The router already has typed signals, decisions, candidate model selection, Router Replay/eval surfaces, and bounded Router Learning.
- Current `main` has no `RADARSelector`, no production query-difficulty model, and no model-budget ability registry.
- Online Router Learning may change only the selected model inside the configured candidate set and `protection` envelope. Recipe and signal changes remain offline.

## Research scope

1. Reproduce the paper formulation on a versioned benchmark before designing runtime APIs.
2. Define model configuration identity precisely: logical model, deployable version, reasoning controls, cost, and evaluation provenance.
3. Train/evaluate query-difficulty and model-budget ability estimates with leakage-safe in-distribution and out-of-distribution splits.
4. Compare static rules, existing selectors, `routing_sampling`, and simple calibrated classifiers at matched quality/cost.
5. Test calibration, abstention, unsupported-domain behavior, new-model onboarding, drift, and adversarial inputs.
6. If evidence warrants integration, expose difficulty as a typed signal or a future `adaptation.strategy` implementation—without adding a legacy standalone selector API.
7. Keep learning/training offline; a production candidate must pass the Router Learning benchmark/protection contract.

## Non-goals

- Do not assume model self-estimated difficulty is trusted ground truth.
- Do not update IRT parameters synchronously from unverified production feedback.
- Do not hard-code “easy/medium/hard” model examples or universal cost savings.
- Do not bypass authorization, safety, residency, context, or configured candidate constraints.

## Acceptance criteria

- [ ] Paper reproduction and repository-specific benchmark are separate and reproducible.
- [ ] Dataset/model/config versions and label provenance are recorded.
- [ ] OOD, calibration, abstention, latency, cost, and quality are reported with simple baselines.
- [ ] New-model onboarding reports uncertainty and required evaluation coverage.
- [ ] Any runtime integration uses current signal/Router Learning contracts and observe mode first.
- [ ] Results state when RADAR does not outperform simpler approaches.

## Validation entrypoint

```bash
make agent-report ENV=cpu CHANGED_FILES="<space-separated changed files>"
```

**Comment by Anya-Gill** [comment]
@liavweiss I see you're assigned to this issue. Would you like to collaborate, I'd love to work on it as well.

**Comment by Whatsonyourmind** [comment]
Worth building in from day one: in P = σ(α(β − θ)), the θ you route on for OOD queries is extrapolated from the embedding regressor, not measured. So θ̂ carries variance, and that variance is largest exactly where you care — novel inputs far from the fitted manifold.

Two consequences:

1. Routing on σ(α(β − θ̂)) is optimistically biased near the easy/hard boundary. σ is concave above 0.5 and convex below, so by Jensen E[σ(α(β − θ))] ≠ σ(α(β − θ̂)) — plugging in the point estimate misjudges exactly the borderline queries an escalation policy hinges on.

2. Cheap fix, no extra model calls: hold out a calibration split, take the difficulty residuals, run split conformal for a distribution-free interval on θ per query. Escalate when its upper bound crosses threshold. You get finite-sample marginal coverage that survives distribution shift — the OOD property you list.

Interval width also doubles as a free abstention signal: when wide, route to a bigger model rather than committing.

**Comment by sanatb187** [comment]
Hi @Xunzhuo, I see this issue is currently unassigned. I’m a first-time contributor to Semantic Router and would love to work on this!

**Comment by Xunzhuo** [comment]
@sanatb187 yes please go ahead with a proposal first before coding. Thanks

**Comment by sanatb187** [comment]
Thanks @Xunzhuo! I’ve gone through the current Semantic Router flow, Router Learning roadmap, and the RADAR paper. My proposed approach is:

1. Start with a paper reproduction on a versioned reasoning benchmark, likely GPQA-Diamond, using the RADAR query-difficulty and model-budget ability formulation.
2. Keep the reproduction separate from Semantic Router runtime changes and record dataset, model, reasoning-budget/configuration, cost, and evaluation provenance.
3. Use leakage-safe ID/OOD splits to evaluate difficulty and ability estimation.
4. Compare RADAR against simpler baselines such as static routing, existing selectors / `routing_sampling`, and a calibrated classifier at matched quality/cost.
5. Report quality, cost, latency, calibration, abstention, and OOD behavior, including cases where RADAR does not outperform simpler methods.
6. If the results justify integration, I’ll then look at exposing difficulty through the existing typed-signal / Router Learning contracts rather than introducing a standalone selector.

I’ll begin with the reproduction/evaluation layer first and share the results before proposing any runtime integration.

**Comment by sanatb187** [comment]
Hi @Xunzhuo 
Quick progress update: I’ve completed the initial RADAR reproduction and evaluation framework, including GPQA support, IRT-based routing, configurable cost metrics, fixed baselines, and vLLM experiments.

I ran two preliminary setups: one locally with Ollama and another with vLLM on a Kaggle T4. RADAR did not meaningfully outperform the fixed frontier. Locally, latency hypervolume was 0.448 versus 0.466, while output-token hypervolume was effectively tied at 0.470 versus 0.469. On Kaggle, latency hypervolume was 0.375 versus 0.422.

I’m expanding the test suite with the paper’s baselines and evaluation metrics. I’ll then run one paper-aligned GPQA setup before reporting back with a recommendation on whether this warrants a Semantic Router implementation.

Repository: https://github.com/sanatb187/RADAR-paper-implementation

**Comment by sanatb187** [comment]
Quick update: I’ve completed a scoped RADAR reproduction and published the implementation here: https://github.com/sanatb187/RADAR-paper-implementation/tree/feat/evaluation-baselines

I tested two setups:

* A local Ollama pilot (64 train / 32 test) where RADAR was worse on latency hypervolume (`−0.0875` linear, `−0.0180` Chebyshev) and effectively tied on output-token cost (`+0.0010`).
* A paper-aligned GPQA split generated on Kaggle T4s: GPQA Main excluding Diamond for training (250 questions), full Diamond for testing (198 questions), Qwen3 4B/8B AWQ, and reasoning budgets of 0/256/1024 tokens.

For the paper-aligned run:

* Fixed frontier: `0.3474`
* RADAR frontier: `0.3310` (`−0.0164`)
* Calibrated-classifier frontier: `0.3525` (`+0.0051`)
* Best fixed accuracy: `41.9%`
* Best RADAR accuracy: `33.8%`
* Oracle accuracy: `68.2%`

The oracle upper bound shows substantial complementarity among the configurations, but the learned IRT model did not generalize: training loss was `0.394`, while test loss was `0.699`. In this setup, RADAR did not outperform the fixed frontier, while a simpler calibrated classifier slightly did.

This is not a full reproduction of the paper’s entire suite: it uses one benchmark, six self-hosted AWQ configurations, and output-token cost instead of provider pricing. Based on this scoped result, I don’t think runtime integration is justified at this stage.

**Comment by Xunzhuo** [comment]
Thanks for documenting the negative result and its limits. Please link the completed run manifests and full evaluator output for the GPQA experiment, pinned to the implementation commit, including the calibration and risk-coverage results. The branch publishes code and headline numbers, but those run outputs are not linked yet; they are the next evidence needed to review the reproduction without starting runtime integration.

**Comment by sanatb187** [comment]
Hi @Xunzhuo,
I’ve added the completed evidence bundle, pinned at commit:
https://github.com/sanatb187/RADAR-paper-implementation/commit/9e0be56b4483151d9dcba1f52de77463e4755be9

Evidence index and provenance:
https://github.com/sanatb187/RADAR-paper-implementation/blob/9e0be56b4483151d9dcba1f52de77463e4755be9/reproductions/gpqa-paper-aligned/README.md

Qwen3-4B AWQ completed manifest:
https://github.com/sanatb187/RADAR-paper-implementation/blob/9e0be56b4483151d9dcba1f52de77463e4755be9/reproductions/gpqa-paper-aligned/manifest_qwen3-4b-awq_budgets-0-256-1024.json

Qwen3-8B AWQ completed manifest:
https://github.com/sanatb187/RADAR-paper-implementation/blob/9e0be56b4483151d9dcba1f52de77463e4755be9/reproductions/gpqa-paper-aligned/manifest_qwen3-8b-awq_budgets-0-256-1024.json

Full evaluator output:
https://github.com/sanatb187/RADAR-paper-implementation/blob/9e0be56b4483151d9dcba1f52de77463e4755be9/reproductions/gpqa-paper-aligned/evaluation_chebyshev_output-tokens.txt

The evaluator output includes the full fixed, RADAR, calibrated-classifier, routing-sampling, and Random-Pair results, along with calibration metrics, per-strategy risk-coverage/AURC values, proxy CPT, and routing diagnostics.

The manifests record clean completed runs and pin generation to commit `04c06897d4e4d36dc941f1b02c4f012b36507ef0`. The merged evaluation implementation is pinned at commit `3ee8ec2d2294216f4c2454990895181af9848336`.

Raw GPQA prompts and responses remain excluded because the dataset is gated.
